Sources/SwiftGRPCNIO/HTTP1ToRawGRPCServerCodec.swift

```diff
@@ -27,111 +27,200 @@ public enum RawGRPCServerResponsePart {
+      throw StateMachineError("expecteded state .expectingBody(_), got \(inboundState)")
+    }
+
+    return .expectingBody(try processBodyState(ctx: ctx, initialState: bodyState, messageBuffer: &body))
+  }
+
+  func processBodyState(ctx: ChannelHandlerContext, initialState: InboundState.Body, messageBuffer: inout ByteBuffer) throws -> InboundState.Body {
+    var bodyState = initialState
+
+    // Iterate over all available incoming data, trying to read length-delimited messages.
+    // Each message has the following format:
+    // - 1 byte "compressed" flag (currently always zero, as we do not support for compression)
+    // - 4 byte signed-integer payload length (N)
+    // - N bytes payload (normally a valid wire-format protocol buffer)
+    while true {
+      switch bodyState {
+      case .expectingCompressedFlag:
+        guard let compressionFlag: Int8 = messageBuffer.readInteger() else { return .expectingCompressedFlag }
+
+        // TODO: Add support for compression.
+        guard compressionFlag == 0 else { throw GRPCStatus.unsupportedCompression }
+
+        bodyState = .expectingMessageLength
+
+      case .expectingMessageLength:
+        guard let messageLength: UInt32 = messageBuffer.readInteger() else { return .expectingMessageLength }
+        bodyState = .receivedMessageLength(messageLength)
+
+      case .receivedMessageLength(let messageLength):
+        // We need to account for messages being spread across multiple `ByteBuffer`s so buffer them
+        // into `buffer`. Note: when messages are contained within a single `ByteBuffer` we're just
+        // taking a slice so don't incur any extra writes.
+        guard messageBuffer.readableBytes >= messageLength else {
+          let remainingBytes = messageLength - numericCast(messageBuffer.readableBytes)
+
+          if var buffer = buffer {
+            buffer.write(buffer: &messageBuffer)
+            self.buffer = buffer
+          } else {
+            messageBuffer.reserveCapacity(numericCast(messageLength))
+            self.buffer = messageBuffer
+          }
+
+          return .receivedMessageLength(remainingBytes)
+        }
+
+        // We know buffer.readableBytes >= messageLength, so it's okay to force unwrap here.
+        var slice = messageBuffer.readSlice(length: numericCast(messageLength))!
+
+        if var buffer = buffer {
+          buffer.write(buffer: &slice)
+          ctx.fireChannelRead(self.wrapInboundOut(.message(buffer)))
+        } else {
+          ctx.fireChannelRead(self.wrapInboundOut(.message(slice)))
+        }
+
+        self.buffer = nil
+        bodyState = .expectingCompressedFlag
       }
-      ctx.fireChannelRead(self.wrapInboundOut(.end))
     }
   }
+
+  private func processEnd(ctx: ChannelHandlerContext, trailers: HTTPHeaders?) throws -> InboundState {
+    guard trailers == nil else {
+      throw StateMachineError("unexpected trailers received \(String(describing: trailers))")
```

Comment:
change suggestion:
```
if let trailers = trailers {
	      throw StateMachineError( "client sent unexpected trailers \(trailers)")
}
```
Avoids printing "Optional(trailers...)" into the string.

Reply:
Good call!